File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int startRow = 0;
        int startCol = 0;
        int endRow = matrix.length-1;
        int endCol = matrix[0].length-1;
        List<Integer> res = new ArrayList<>();
        while(startRow <= endRow && startCol <= endCol) {
            
            for(int i=startCol; i<=endCol; i++) {
                res.add(matrix[startRow][i]);
            }
            
            startRow++;
            
            for(int i=startRow; i<=endRow; i++) {
                res.add(matrix[i][endCol]);
            }
            
            endCol--;
            
            if(startRow-1 != endRow) {
                for(int i=endCol; i>=startCol; i--) {
                    res.add(matrix[endRow][i]);
                }
            }
            
            endRow--;
            
            if(startCol != endCol+1) {
                for(int i=endRow; i>=startRow; i--) {
                    res.add(matrix[i][startCol]);
                }
            }
            
            startCol++;
        }
        
        return res;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        m = 0
        n = 0
        for i in range(min(math.ceil(len(matrix)/2), math.ceil(len(matrix[0])/2))):
            
            for j in range(n, len(matrix[0])-n):
                res.append(matrix[m][j])
            
            for j in range(m+1, len(matrix)-m):
                res.append(matrix[j][len(matrix[0])-n-1])
            
            if m != len(matrix)-m-1:
                for j in reversed(range(n, len(matrix[0])-n-1)):
                    res.append(matrix[len(matrix)-m-1][j])
            
            if n != len(matrix[0])-n-1:
                for j in reversed(range(m+1, len(matrix)-m-1)):
                    res.append(matrix[j][n])
                
            m += 1
            n += 1
                
        return res